sdk/core/core-rest-pipeline/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -32,26 +63,90 @@ export interface BearerTokenAuthenticationPolicyOptions {
+  const { prepareRequest, processChallenge } = challengeCallbacks ?? {};
+
+  /**
+   * retrieveToken will call to the underlying credential's getToken request with properties coming from the request,
+   * as well as properties coming from parsing the CAE challenge, if any.
+   */
+  async function retrieveToken(
+    request: PipelineRequest,
+    context?: BearerTokenChallengeResult
+  ): Promise<string | undefined> {
     let accessToken = tokenCache.getCachedToken();
+    const getTokenOptions: GetTokenOptions = {
+      claims: context?.claims,
+      abortSignal: request.abortSignal,
+      tracingOptions: request.tracingOptions
+    };
     if (accessToken === undefined) {
-      accessToken = (await credential.getToken(scopes, tokenOptions)) || undefined;
+      accessToken =
+        (await credential.getToken(context?.scopes || scopes, getTokenOptions)) || undefined;
       tokenCache.setCachedToken(accessToken);
     }
-
     return accessToken ? accessToken.token : undefined;
   }
+
+  /**
+   * Populates the token in the request headers.
+   */
+  function assignToken(request: PipelineRequest, token?: string): PipelineRequest {
+    if (token) {
+      request.headers.set("Authorization", `Bearer ${token}`);
+    }
+    return request;
+  }
+
+  /**
+   * We will retrieve the challenge only if the response status code was 401,
+   * and if the response contained the header "WWW-Authenticate" with a non-empty value.
+   */
+  function getChallenge(response: PipelineResponse): string | undefined {
+    const challenge = response.headers.get("WWW-Authenticate");
+    if (response.status === 401 && challenge) {
+      return challenge;
+    }
+    return;
+  }
+
   return {
     name: bearerTokenAuthenticationPolicyName,
+    /**
+     * If there's no challenge parameter:
+     * - It will try to retrieve the token using the cache, or the credential's getToken.
+     * - Then it will try the next policy with or without the retrieved token.
+     *
+     * It uses the challenge parameters to:
+     * - Skip a first attempt to get the token from the credential if there's no cached token,
+     *   since it expects the token to be retrievable only after the challenge.
+     * - Prepare the outgoing request if the `prepareRequest` method has been provided.
+     * - Send an initial request to receive the challenge if it fails.
+     * - Process a challenge if the response contains it.
+     * - Retrieve a token with the challenge information, then re-send the request.
+     */
     async sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse> {
-      const token = await getToken({
-        abortSignal: request.abortSignal,
-        tracingOptions: request.tracingOptions
-      });
-      request.headers.set("Authorization", `Bearer ${token}`);
-      return next(request);
+      const accessToken = tokenCache.getCachedToken();
+      if (!accessToken && !processChallenge) {
```

Comment:
I'm confused about something here, for the scenario where a pipeline isn't using a challenge. In that case `processChallenge` should always be falsy. 

This if condition means we won't try to retrieve the token initially, which will make the request fail as unauthenticated, but then the below `if (challenge && processChallenge)` will make it so we don't retry either. 

So wouldn't every authenticated request fail when `processChallenge` isn't passed?

Reply:
oh wait, the not defined makes this true! Ahhhh I'm in need of caffeine. Now I get it

Reply:
I just had to click send review and embarrass myself first 😂